pallet-parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,557 @@
+			if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
+				let relay_chain_height = Self::validation_function_params().relay_chain_height;
+				if relay_chain_height >= apply_block {
+					PendingValidationFunction::kill();
+					Self::put_parachain_code(&validation_function);
+					Self::deposit_event(Event::ValidationFunctionApplied(relay_chain_height));
+				}
+			}
+		}
+
+		pub fn set_code(origin, validation_function: ValidationFunction) {
+			// TODO: in the future, we can't rely on a superuser existing
+			// on-chain who can just wave their hands and make this happen.
+			// Instead, this should hook into the democracy pallet and check
+			// that a validation function upgrade has been approved; potentially,
+			// it should even trigger the validation function upgrade automatically
+			// the moment the vote passes.
+			System::<T>::can_set_code(origin, &validation_function)?;
+			ensure!(!PendingValidationFunction::exists(), Error::<T>::OverlappingUpgrades);
+			let vfp = Self::validation_function_params();
+			ensure!(validation_function.len() <= vfp.max_code_size as usize, Error::<T>::TooBig);
+			let apply_block = vfp.code_upgrade_allowed.ok_or(Error::<T>::ProhibitedByPolkadot)?;
+
+			// When a code upgrade is scheduled, it has to be applied in two
+			// places, synchronized: both polkadot and the individual parachain
+			// have to upgrade on the same relay chain block.
+			//
+			// `put_polkadot_code` notifies polkadot; the `PendingValidationFunction`
+			// storage keeps track locally for the parachain upgrade, which will
+			// be applied later.
+			Self::put_polkadot_code(&validation_function);
+			PendingValidationFunction::put((apply_block, validation_function));
+			Self::deposit_event(Event::ValidationFunctionStored(apply_block));
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get validation function parameters.
+	///
+	/// This tries to get them first from a magic storage key which is injected
+	/// by cumulus; if that value is set, we should use it, because it's more
+	/// trustworthy than a block's contents. That said, we also try to extract
+	/// them from this block's extrinsics; cumulus also injects them as an
+	/// inherent into each block, so that they're available during block production.
+	fn validation_function_params() -> ValidationFunctionParams {
+		// this storage value is set by cumulus during block validation
+		storage::unhashed::get(VALIDATION_FUNCTION_PARAMS)
+			// this storage value is set each block by the ProvideInherent
+			// implementation, ensuring that the data is also available
+			// during block production
+			.unwrap_or_else(|| VFPs::get())
+			// in the event that something has gone seriously wrong and neither
+			// of those values are present, I believe that VFPs::get falls back
+			// to the Default::default() value, which should be harmless in that
+			// it prevents users from doing any upgrades during the invalid block.
```

Comment:
Assumption about what happens in the event that the VFPs are not initialized for whatever reason. The most likely causes of this error are that this pallet is used outside a cumulus parachain, or is not included in the runtime's inherent providers.

Note that this doesn't panic. An alternate approach here might be to do `.or_else(|| VFPs::try_get().ok()).expect("VFPs were not set; is the runtime configured properly?")`, which would panic in that case. I decided not to, on the theory that panics are bad and should be avoided, but given that this is caused by misconfiguration, it might be desirable to do that instead.